### PR TITLE
[Hack] Add logic to BuildConfig to allow for announcement debugging.

### PR DESCRIPTION
**Pull request by notandyvee**
The current announcement logic is a bit of a pain to test. Normally I'd add a feature flag. But I thought instead let's add something in build.gradle to use local.properties so it's a bit easier to test.

Now you can simply add `alwaysShowAnnouncement=true` to your `local.properties`, build the app in debug, and the announcements from your current release should be shown every time you load the app.

Thoughts?

-----

## To Test:

- [ ] Go to the Me tab
- [ ] Debug Settings
- [ ] Press the overflow menu and click `All shard preference flag`.
- [ ] You should see `prefs_always_show_announcement`. Make sure it's checked.
- [ ] When you exit the debug settings screen you will begin to see the announcement pop up often.

**Note**: I was going to add documentation, but instead decided against it until this PR is reviewed.

-----

## Regression Notes

1. Potential unintended areas of impact

    n/a

2. What I did to test those areas of impact (or what existing automated tests I relied on)

    Manual test.

3. What automated tests I added (or what prevented me from doing so)

    n/a

-----

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by guarani** [comment]
👋 My setup wasn't working so I couldn't test. See https://a8c.slack.com/archives/C441E3YTS/p1711413127791989?thread_ts=1711405412.749429&cid=C441E3YTS for details.

Given this, I think it would be best if @irfano could review 🙏 , so I removed myself as a reviewer.

**Comment by irfano** [comment]
@notandyvee, I am the reviewer but the PR is still draft. Is this ready for review?

**Comment by notandyvee** [comment]
Apologies @irfano I just haven't had time. Let me carve out time today to make the documentation and then convert it to a final PR. Since Paul was having trouble it became clear it needs some documentation.

**Comment by notandyvee** [comment]
It's ready @irfano . I held off on writing some docs until I hear what you think. Please note this PR does not need to be merged. If you feel strongly against adding this, we can close and consider a different way of accomplishing the same thing. No pressure. Adding @aditi-bhatia to the reviewers to get some more opinions.

**Review by aditi-bhatia — APPROVED** [review]
Looks good to me, this is useful to have added in 👍 I was unable to test the `extra credit` with the release build because of some issues, but all other steps worked as expected. I'll try the release build again tomorrow but not expecting any issues with that

**Comment by irfano** [comment]
I added `alwaysShowAnnouncement=true` to `local.properties` and built `jetpackWasabiDebug` but couldn't see the announcement. What am I missing? 🤔

**Comment by notandyvee** [comment]
> I added alwaysShowAnnouncement=true to local.properties and built jetpackWasabiDebug but couldn't see the announcement. What am I missing? 🤔

I don't use wasabi. This could be another area of concern. Could you try jalapeno first to see whatsup @irfano ?

**Comment by irfano** [comment]
Our purpose of this is to see how the announcement will look on the screen, right? I find putting a config in the Debug settings screen preferable because testers who are not Android developers can also utilize this setting. WDYT?

I noticed that iOS has some buttons for similar features on its Debug settings screen. We can do the same.

**Comment by notandyvee** [comment]
> I noticed that iOS has some buttons for similar features on its Debug settings screen. We can do the same.

I'll need some time to look into how iOS is doing it. Android's is a bit clunky, which is why I decided against it. But sure.

**Comment by irfano** [comment]
We have a "FEATURES IN DEVELOPMENT" tab in the "Debug settings". Perhaps we can add a fourth tab to this screen, "LOCAL CONFIGS", and use the same design as "FEATURES IN DEVELOPMENT.

I believe being able to find everything related to testing configs within "Debug settings" would be a better option.

**Comment by ravishanker** [comment]
It appeared to me on **trunk** build once.  Is that the upgrade to 24.7 scenario showing up?

**Comment by ravishanker** [comment]
> We have a "FEATURES IN DEVELOPMENT" tab in the "Debug settings". Perhaps we can add a fourth tab to this screen, "LOCAL CONFIGS", and use the same design as "FEATURES IN DEVELOPMENT.
> 
> I believe being able to find everything related to testing configs within "Debug settings" would be a better option.
> 


You could just add it to Features in development tab, and stays there forever.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          2 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>SharedPrefsField</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>SharedPrefsBooleanField</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>strings.xml</code> files should only be updated on release branches, when the translations are downloaded by our automation.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by notandyvee** [comment]
@irfano sorry I didn't get to this sooner. Mind taking a look when you get a chance and give me your thoughts. I did what was mentioned. Add a new tab on the debug settings screen. I decided to create a new `UiItem`, `SharedPrefsField`. This item currently only represents a boolean, but can be a string, long, etc. Hence why I am using a type. I tried to keep it simple since it's only for debugging.

**Comment by irfano** [comment]
@notandyvee, while reviewing the change, I noticed a screen that already contains your new pref, `pref_always_show_announcement`. This screen was introduced 4 months ago (#19880). Do you think we can just add a new pref and use this existing screen?

**Comment by notandyvee** [comment]
> @notandyvee, while reviewing the change, I noticed a screen that already contains your new pref, pref_always_show_announcement. This screen was introduced 4 months ago (https://github.com/wordpress-mobile/WordPress-Android/pull/19880). Do you think we can just add a new pref and use this existing screen?

Sigh 🤦🏻‍♂️ . I do think we should re-use that screen. Just sucks we didn't catch that sooner. Gonna be the 3rd iteration. I'll look into it on monday.

**Comment by notandyvee** [comment]
@justtwago I made some edits to `DebugSharedPreferenceFlagsViewModel` that require your approval. I needed a debug shared pref value for debugging. Since the shared pref needs to be written first before it's seen on that screen, I thought it made more sense to add any flags we want to test in a list of enums. This way any test ones can be toggled.

The other big reason I felt this change would be helpful is to open it up for strings and other primitive types. Maybe test urls or other text we can change for debug purposed. I didn't make that change as it's not required yet, but it's why I added `KClass` in the `DebugPrefs` enum. Thoughts?

File: WordPress/src/main/java/org/wordpress/android/ui/debug/preferences/DebugPrefs.kt
Comment (irfano) [inline WordPress/src/main/java/org/wordpress/android/ui/debug/preferences/DebugPrefs.kt]: What does "Class used to track" mean?

**Review by irfano — CHANGES_REQUESTED** [review]
It's a great solution Andy, I like it. Adding some notes:

- `DebugSharedPreferenceFlagsViewModelTest` is failing.
- Now, I can only see `pres_always_show_announcement`.

|before|after|
|-|-|
|||

**Comment by notandyvee** [comment]
@irfano I'll fix the failing test.

> What does "Class used to track" mean?

I started to add class doc and stopped mid way 😅 . Updated now.

> Now, I can only see pres_always_show_announcement

This is on purpose. It's why I pinged @justtwago . Currently the preference flag needs to exist in order for it to show up. I just updated it to be more explicit. If you have no issues feel free to approve, but **do not merge**. We can wait for Arty to put his opinion.

**Review by irfano — APPROVED** [review]

**Comment by justtwago** [comment]
Really nice idea, Andy! Thanks for pinging me!
I'm sorry that the debug screen I introduced wasn't clearly described. The initial idea was to display all the locally stored flags so we could test/reset them without clearing the cache/local data. My motivation was the "Dynamic dashboard cards" project (PCYsg-W09-p2). Once we tap on "Hide" from a card's menu, it creates a shared pref for this particular card to stop displaying it on the dashboard. The only way to reset or "unhide" it was by clearing the cache and going through the authentication process again.

What if we combined the two lists to display some app-managed flags and the predefined debug flags that you introduced?

File: WordPress/src/main/java/org/wordpress/android/ui/debug/preferences/DebugSharedPreferenceFlagsViewModel.kt
Comment (justtwago) [inline WordPress/src/main/java/org/wordpress/android/ui/debug/preferences/DebugSharedPreferenceFlagsViewModel.kt]: Really nice idea, Andy! Thanks for pinging me!
I'm sorry that the debug screen I introduced wasn't clearly described. The initial idea was to display all the locally stored flags so we could test/reset them without clearing the cache/local data. My motivation was the "Dynamic dashboard cards" project (PCYsg-W09-p2). Once we tap on "Hide" from a card's menu, it creates a shared pref for this particular card to stop displaying it on the dashboard. The only way to reset or "unhide" it was by clearing the cache and going through the authentication process again.

What if we combined the two lists to display app-managed flags and the predefined debug flags that you introduced?

Comment (justtwago) [reply WordPress/src/main/java/org/wordpress/android/ui/debug/preferences/DebugSharedPreferenceFlagsViewModel.kt]: This is the context: pcdRpT-5f2-p2
I should've linked it to my PR 🤦‍♂️

Comment (notandyvee) [reply WordPress/src/main/java/org/wordpress/android/ui/debug/preferences/DebugSharedPreferenceFlagsViewModel.kt]: > What if we combined the two lists to display some app-managed flags and the predefined debug flags that you introduced?

I went ahead and merged the two maps. Would be great if you could do a review @justtwago 🙏🏻

**Review by justtwago — APPROVED** [review]
Thanks so much for the improvement! LG2M! 🎸

**Pull request merged**
